contracts/cw-proposal-multiple/src/proposal.rs

```diff
@@ -0,0 +1,177 @@
+    /// proposal. None if no deposit.
+    pub deposit_info: Option<CheckedDepositInfo>,
+}
+
+impl Proposal for MultipleChoiceProposal {
+    fn proposer(&self) -> Addr {
+        self.proposer.clone()
+    }
+    fn deposit_info(&self) -> Option<CheckedDepositInfo> {
+        self.deposit_info.clone()
+    }
+    fn status(&self) -> Status {
+        self.status
+    }
+}
+
+impl MultipleChoiceProposal {
+    /// Consumes the proposal and returns a version which may be used
+    /// in a query response. The difference being that proposal
+    /// statuses are only updated on vote, execute, and close
+    /// events. It is possible though that since a vote has occured
+    /// the proposal expiring has changed its status. This method
+    /// recomputes the status so that queries get accurate
+    /// information.
+    pub fn into_response(mut self, block: &BlockInfo, id: u64) -> ProposalResponse {
+        self.update_status(block);
+        ProposalResponse { id, proposal: self }
+    }
+
+    /// Gets the current status of the proposal.
+    pub fn current_status(&self, block: &BlockInfo) -> Status {
+        if self.status == Status::Open && self.is_passed() {
+            Status::Passed
+        } else if self.status == Status::Open
+            && (self.expiration.is_expired(block) || self.is_rejected(block))
+        {
+            Status::Rejected
+        } else {
+            self.status
+        }
+    }
+
+    /// Sets a proposals status to its current status.
+    pub fn update_status(&mut self, block: &BlockInfo) {
+        self.status = self.current_status(block);
+    }
+
+    /// Returns true iff this proposal is sure to pass (even before
+    /// expiration if no future sequence of possible votes can cause
+    /// it to fail). Passing in the case of multiple choice proposals
+    /// means that one of the options that is not "None of the above" has won the most votes.
+    pub fn is_passed(&self) -> bool {
+        // Proposal can only pass if quorum has been met.
+        if does_vote_count_pass(
+            self.votes.total(),
+            self.total_power,
+            self.voting_strategy.get_quorum(),
+        ) {
+            let (choice_idx, winning_choice) = self.calculate_winning_choice();
+            // Check that the winning choice is not None.
+            if winning_choice.option_type != MultipleChoiceOptionType::None {
+                // If the leading choice cannot possibly be outwon by the "None" option, the proposal has passed.
+                // This means that the difference between them must be greater than the remaining vote power.
+                let none_count = u128::from(self.get_none_vote_count());
+                let winning_choice_count = self.votes.vote_weights[choice_idx].u128();
+                let remaining_power = self.total_power.sub(self.votes.total()).u128();
+                if winning_choice_count - none_count > remaining_power {
```

Comment:
```
total_power - 7
foo - 3
bar - 2
none - 1
```

Here `winning_choice_count - none_count > remaining_power` but it is still possible for bar to get the next vote and result in a tie. Is this where the 'first across the line' logic comes in?

Reply:
hmm fair point the tie case needs to be handled as well